File: probs/2096.py
import sys
input = sys.stdin.readline

input = open('input.txt', 'r').readline

N = int(input())

maxdp = [[0]*3 for _ in range(2)]
mindp = [[0]*3 for _ in range(2)]
b = list(map(int, input().split()))
maxdp[0] = b[:]
mindp[0] = b[:]
for i in range(1, N):
    b = list(map(int, input().split()))
    maxdp[i%2][0] = max(
        b[0]+maxdp[(i+1)%2][0],
        b[0]+maxdp[(i+1)%2][1]
    )
    maxdp[i%2][1] = max(
        b[1]+maxdp[(i+1)%2][0],
        b[1]+maxdp[(i+1)%2][1],
        b[1]+maxdp[(i+1)%2][2],
    )
    maxdp[i%2][2] = max(
        b[2]+maxdp[(i+1)%2][1],
        b[2]+maxdp[(i+1)%2][2]
    )

    mindp[i%2][0] = min(
        b[0]+mindp[(i+1)%2][0],
        b[0]+mindp[(i+1)%2][1]
    )
    mindp[i%2][1] = min(
        b[1]+mindp[(i+1)%2][0],
        b[1]+mindp[(i+1)%2][1],
        b[1]+mindp[(i+1)%2][2],
    )
    mindp[i%2][2] = min(
        b[2]+mindp[(i+1)%2][1],
        b[2]+mindp[(i+1)%2][2]
    )

print(
    max(maxdp[(N+1)%2]),
    min(mindp[(N+1)%2])
)

File: probs/2096.cpp
#include <bits/stdc++.h>

using namespace std;

const int INF = 1e9;

int dp[2][3][2];

int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    freopen("input.txt", "r", stdin);

    int N; cin >> N;

    for (int i = 0; i < 3; i++) {
        cin >> dp[0][i][0];
        dp[0][i][1] = dp[0][i][0];
    }
    for (int i = 1; i < N; i++) {
        int ipt[3];
        for (int j = 0; j < 3; j++) cin >> ipt[j];

        dp[i%2][0][0] = ipt[0] + max({
            dp[(i+1)%2][0][0],
            dp[(i+1)%2][1][0],
        });
        dp[i%2][1][0] = ipt[1] + max({
            dp[(i+1)%2][0][0],
            dp[(i+1)%2][1][0],
            dp[(i+1)%2][2][0],
        });
        dp[i%2][2][0] = ipt[2] + max({
            dp[(i+1)%2][1][0],
            dp[(i+1)%2][2][0],
        });

        dp[i%2][0][1] = ipt[0] + min({
            dp[(i+1)%2][0][1],
            dp[(i+1)%2][1][1],
        });
        dp[i%2][1][1] = ipt[1] + min({
            dp[(i+1)%2][0][1],
            dp[(i+1)%2][1][1],
            dp[(i+1)%2][2][1],
        });
        dp[i%2][2][1] = ipt[2] + min({
            dp[(i+1)%2][1][1],
            dp[(i+1)%2][2][1],
        });
    }

    cout << max({
        dp[(N+1)%2][0][0],
        dp[(N+1)%2][1][0],
        dp[(N+1)%2][2][0],
    }) << ' ' << min({
        dp[(N+1)%2][0][1],
        dp[(N+1)%2][1][1],
        dp[(N+1)%2][2][1],
    }) << '\n';

    return 0;
}
